mm2src/coins/solana.rs

```diff
@@ -378,6 +396,246 @@ impl SolanaCoin {
+                "Provided secret_hash {:?} does not match transaction secret_hash {:?}",
+                secret_hash, tx_secret_hash
+            ));
+        }
+        let (vault_pda, vault_pda_data, vault_bump_seed, vault_bump_seed_data, _rent_exemption_lamports) =
+            try_tx_fus!(self.create_vaults(lock_time, secret_hash, swap_program_id, PAYMENT_STORAGE_SPACE));
+        let swap_instruction = AtomicSwapInstruction::ReceiverSpend {
+            secret,
+            lock_time,
+            amount,
+            sender,
+            token_program,
+            vault_bump_seed,
+            vault_bump_seed_data,
+        };
+        let accounts = vec![
+            AccountMeta::new(self.key_pair.pubkey(), true),
+            AccountMeta::new(vault_pda_data, false),
+            AccountMeta::new(vault_pda, false),
+            AccountMeta::new(solana_program::system_program::id(), false),
+        ];
+        self.sign_and_send_transaction(swap_program_id, accounts, swap_instruction.pack())
+    }
+
+    fn refund_hash_time_locked_payment(&self, args: RefundPaymentArgs) -> SolTxFut {
+        let receiver = Pubkey::new(args.other_pubkey.iter().as_slice());
+        let swap_program_id = Pubkey::new(
+            try_tx_fus_opt!(
+                args.swap_contract_address.as_ref(),
+                format!(
+                    "Unable to extract Bytes from args.swap_contract_address ( {:?} )",
+                    args.swap_contract_address
+                )
+            )
+            .as_slice(),
+        );
+        let (lock_time, secret_hash, amount, token_program) =
+            try_tx_fus!(self.get_transaction_details(args.payment_tx));
+        let (vault_pda, vault_pda_data, vault_bump_seed, vault_bump_seed_data, _rent_exemption_lamports) =
+            try_tx_fus!(self.create_vaults(lock_time, secret_hash, swap_program_id, PAYMENT_STORAGE_SPACE));
+        let swap_instruction = AtomicSwapInstruction::SenderRefund {
+            secret_hash,
+            lock_time,
+            amount,
+            receiver,
+            token_program,
+            vault_bump_seed,
+            vault_bump_seed_data,
+        };
+        let accounts = vec![
+            AccountMeta::new(self.key_pair.pubkey(), true), // Marked as signer
+            AccountMeta::new(vault_pda_data, false),        // Not a signer
+            AccountMeta::new(vault_pda, false),             // Not a signer
+            AccountMeta::new(solana_program::system_program::id(), false), //system_program must be included
+        ];
+        self.sign_and_send_transaction(swap_program_id, accounts, swap_instruction.pack())
+    }
+
+    fn get_transaction_details(&self, tx_hash: &[u8]) -> Result<(u64, [u8; 32], u64, Pubkey), Box<TransactionErr>> {
+        let transaction = deserialize(tx_hash);
+        let transaction: SolTransaction =
+            transaction.map_err(|e| Box::new(TransactionErr::Plain(ERRL!("error deserializing tx_hash: {:?}", e))))?;
```

Comment:
In one line is better
```suggestion
        let transaction: SolTransaction = deserialize(tx_hash)
            .map_err(|e| Box::new(TransactionErr::Plain(ERRL!("error deserializing tx_hash: {:?}", e))))?;
```